runtime/lua/vim/lsp/diagnostic.lua

```diff
@@ -1196,42 +1196,62 @@ end
 ---             - Exclusive severity to consider. Overrides {severity_limit}
 ---         - {severity_limit}: (DiagnosticSeverity)
 ---             - Limit severity of diagnostics found. E.g. "Warning" means { "Error", "Warning" } will be valid.
+---         - {workspace}: (boolean, default false)
+---             - Set the list with workspace diagnostics
 function M.set_loclist(opts)
   opts = opts or {}
 
   local open_loclist = if_nil(opts.open_loclist, true)
 
-  local bufnr = vim.api.nvim_get_current_buf()
-  local buffer_diags = M.get(bufnr, opts.client_id)
-
-  if opts.severity then
-    buffer_diags = filter_to_severity_limit(opts.severity, buffer_diags)
-  elseif opts.severity_limit then
-    buffer_diags = filter_by_severity_limit(opts.severity_limit, buffer_diags)
-  end
+  local current_bufnr = vim.api.nvim_get_current_buf()
+  local diags = opts.workspace and M.get_all() or {
+    [current_bufnr] = M.get(current_bufnr, opts.client_id)
+  }
 
   local items = {}
-  local insert_diag = function(diag)
+  local insert_diag = function(bufnr, diag)
     local pos = diag.range.start
     local row = pos.line
-    local col = util.character_offset(bufnr, row, pos.character)
 
-    local line = (api.nvim_buf_get_lines(bufnr, row, row + 1, false) or {""})[1]
+    local line, col, text
+    if api.nvim_buf_is_loaded(bufnr) then
+      line = (api.nvim_buf_get_lines(bufnr, row, row + 1, false) or {""})[1]
+      col = util.character_offset(bufnr, row, pos.character) or 0
```

Comment:
Not really relevant to the changes in this PR as it only moves the logic, but `character_offset` here is wrong, we'd need to convert the utf-16 offset into a byteindex using `vim.str_byteindex()` - and if `client.offset_encoding` is `utf-8` we can use `pos.character` as is without conversion.

Reply:
at that point, there is no information which clients produced which diagnostic entries. Such a change would need to happen elsewhere